cmds/chroot/chroot.go

```diff
@@ -0,0 +1,206 @@
+		"inside the chroot. This option is only permitted when newroot is the old / directory."))
+	flag.Parse()
+}
+
+func stringToUint32(str string) (uint32, error) {
+	ret, err := strconv.ParseUint(str, 10, 32)
+	if err != nil {
+		return 0, err
+	}
+	return uint32(ret), nil
+}
+
+func getCurrentUser() (uid uint32, gid uint32) {
+	return uint32(syscall.Geteuid()), uint32(syscall.Getegid())
+}
+
+func parseUserspec() (uid uint32, gid uint32, err error) {
+	userspecSplit := strings.Split(userspecFlag, ":")
+	if len(userspecSplit) != 2 || userspecSplit[1] == "" {
+		return 0, 0, fmt.Errorf("Syntax error")
+	}
+
+	uid, err = stringToUint32(userspecSplit[0])
+	if err != nil {
+		return 0, 0, err
+	}
+
+	gid, err = stringToUint32(userspecSplit[1])
+	if err != nil {
+		return 0, 0, err
+	}
+
+	return uid, gid, nil
+}
+
+func parseGroups() (groups []uint32, err error) {
+	groupStrs := strings.Split(groupsFlag, ",")
+	ret := make([]uint32, len(groupStrs))
+	for index, group := range groupStrs {
+
+		gid, err := stringToUint32(group)
+		if err != nil {
+			return ret, err
+		}
+
+		ret[index] = gid
+	}
+	return ret, nil
+}
+
+func parseCommand(args []string) []string {
+	if len(args) > 1 {
+		return args[1:]
+	}
+	return []string{"/bin/sh", "-i"}
+}
+
+func parseRoot(args []string) (root string, err error) {
+	if len(args) < 1 {
+		return "", fmt.Errorf("Missing operand")
+	}
+
+	root, err = filepath.Abs(args[0])
+	if err != nil {
+		return "", err
+	}
+
+	return root, nil
+}
+
+func isRoot(dir string) (bool, error) {
+	realPath, err := filepath.EvalSymlinks(dir)
+	if err != nil {
+		return false, err
+	}
+	absolutePath, err := filepath.Abs(realPath)
+	if err != nil {
+		return false, err
+	}
+	if absolutePath == "/" {
+		return true, nil
+	}
+	return false, nil
+}
+
+func setIDs(uid *uint32, gid *uint32, extraGids *[]uint32) (err error) {
+	var (
+		tuid       uint32
+		tgid       uint32
+		textraGids []uint32
+	)
+
+	if userspecFlag == "" {
+		tuid, tgid = getCurrentUser()
+	} else {
+		tuid, tgid, err = parseUserspec()
+		if err != nil {
+			return err
+		}
+	}
+
+	if groupsFlag != "" {
+		textraGids, err = parseGroups()
+		if err != nil {
+			return err
+		}
+	}
+
+	*uid = tuid
+	*gid = tgid
+	*extraGids = textraGids
+	return nil
+}
+
+func die(err error) {
+	fmt.Println(err)
+	os.Exit(1)
+}
+
+func main() {
+	var (
+		uid       uint32
+		gid       uint32
+		extraGids []uint32
+		newRoot   string
+		isOldroot bool
+		err       error
+	)
+
+	if flag.NFlag() == 0 && flag.NArg() == 0 {
+		flag.PrintDefaults()
+		os.Exit(1)
+	}
+
+	newRoot, err = parseRoot(flag.Args())
+	if err != nil {
+		die(err)
+	}
+	isOldroot, err = isRoot(newRoot)
+	if err != nil {
+		die(err)
+	}
+
+	if !skipchdirFlag {
+		err = os.Chdir(newRoot)
+		if err != nil {
+			die(err)
+		}
+	} else if !isOldroot {
+		die(fmt.Errorf("The -s option is only permitted when newroot is the old / directory"))
+	}
+
+	if err = syscall.Chroot(newRoot); err != nil {
```

Comment:
no, you definitely can not use chroot this way. There's no guarantee that the process that does the chroot is the one that will run the command. You've to to use the Chroot parts of cmd and restart yourself. 

See how pflask does it.

Reply:
This is fine, actually. Go threads share an FS context: http://cs/github/golang/go/src/runtime/os_linux.go?l=131&rcl=9745eed4fd4160cfbf55e9dbbfa99aca5563b392

But, actually, I think that this call is *unnecessary*. You could just set the Chroot attribute in the SysProcAttr struct for exec.Command below. https://golang.org/pkg/syscall/#SysProcAttr

Reply:
This is fine, actually. Go threads share an FS context: https://github.com/golang/go/blob/master/src/runtime/os_linux.go#L131

But, actually, I think that this call is unnecessary. You could just set the Chroot attribute in the SysProcAttr struct for exec.Command below. https://golang.org/pkg/syscall/#SysProcAttr

Reply:
(no need to restart yourself either -- since you're running a command anyway, just set the chroot on that.)

Reply:
well, if it's fine, then something has changed. It used not to be fine. But in any event you've got to to use the Chroot attribute, not the system call.